### PR TITLE
Add resubmission function

### PR DESCRIPTION
**Pull request by armoj**
* Add `/api/v1/submissions/resubmit-failed` endpoint.
* Add `getQuestionnaireIdsBySubmissionStatus` DAL function.
* Add tests

Tests pass.

Manually tested using Postman to hit the new endpoint. Emulated `FAILED` submission via row updates using PG Admin.

### CODE REVIEW

File: questionnaire/questionnaire-service.js
Comment (armoj) [reply questionnaire/questionnaire-service.js:595]: This is the reason I wanted to use the other functionality we chatted about. What I could do here is update the submission to IN_PROGRESS at this point and it would avoid that issue.

File: questionnaire/submissions.js
Comment (armoj) [reply questionnaire/submissions.js:18]: only if we are in favour of not caring about the response

File: questionnaire/submissions-service.js
Comment (armoj) [inline questionnaire/submissions-service.js:24]: should this be await? I am thinking it should. having this be set to IN_PROGRESS first will avoid it being picked up by a subsequent request.

Comment (sinclairs) [reply questionnaire/submissions-service.js:24]: It looks like this could get stuck "IN_PROGRESS" should the message bus call fail

**Review by sinclairs — CHANGES_REQUESTED** [review]

File: questionnaire/submissions-service.js
Comment (BarryPiccinni) [reply questionnaire/submissions-service.js:24]: I'm trying to follow the flow for this, if  the application service was down, I understand they would queue up and be processed eventually, but if the message bus was down would these promises not just return an error, leaving these cases `IN_PROGRESS` and unable to get picked up by a later job?

File: questionnaire/submissions-routes.js
Comment (BarryPiccinni) [inline questionnaire/submissions-routes.js:18]: Given that we now have a distinct submission function and a resubmission function and we shouldn't be able to fall into any unwanted/unexpected statuses, could we not just lock this to only find failed applications? Or are you thinking this still needs to be reusable for other statuses?

**Review by BarryPiccinni — CHANGES_REQUESTED** [review]
As well as getting one failing test locally, I'm also getting quite a number of threshold violations that wont pass the CI pipeline test.

File: questionnaire/submissions-service.js
Comment (armoj) [reply questionnaire/submissions-service.js:24]: are you basically saying that we probably need a try/catch safety net to avoid that situation? Something like the below.

````
const promises = questionnaireIds.map(async id => {
                await updateQuestionnaireSubmissionStatus(id, 'IN_PROGRESS');
                let promise;
                try {
                    promise = messageBusCaller.post('SubmissionQueue', {
                        applicationId: id
                    });
                } catch (err) {
                    await updateQuestionnaireSubmissionStatus(id, 'FAILED');
                }
                return promise;
            });
````

File: questionnaire/submissions.test.js
Comment (sinclairs) [inline questionnaire/submissions.test.js:59]: This doesn't match the message bus caller behaviour

File: questionnaire/submissions.test.js
Comment (sinclairs) [inline questionnaire/submissions.test.js:536]: This test isn't accurate due to the invalid mock

File: questionnaire/submissions-service.js
Comment (sinclairs) [inline questionnaire/submissions-service.js:29]: This won't be hit

File: questionnaire/submissions-service.js
Comment (sinclairs) [inline questionnaire/submissions-service.js:21]: What's the intent for these promises vs fire and forget?

**Review by sinclairs — CHANGES_REQUESTED** [review]
Appears to be a fundamental issue with the logic. I can review the rest when the comments have been addressed.

File: questionnaire/submissions-service.js
Comment (armoj) [reply questionnaire/submissions-service.js:21]: so that it is somewhat testable and so the coverage is not 0%

File: questionnaire/submissions-service.js
Comment (armoj) [reply questionnaire/submissions-service.js:29]: coverage says it is

File: questionnaire/submissions.test.js
Comment (armoj) [reply questionnaire/submissions.test.js:59]: again.. rubbish tests for a quick and dirty endpoint. this `throw` is actually emulating the `got` module throwing. as far as the submissions-service is concerned, the behaviour is identical. Not 100% ideal... however let's get something out. it effectively does exactly as expected.

Comment (sinclairs) [reply questionnaire/submissions.test.js:59]: Either I'm missing something? or this isn't doing what you think it is. `got` is asynchronous and therefore so is the message bus caller. This mock isn't asynchronous. I'm happy to get something out, but it needs to work as expected and I don't think this does.

File: questionnaire/submissions-service.js
Comment (sinclairs) [reply questionnaire/submissions-service.js:29]: I assume that's because the test is wrong? https://github.com/CriminalInjuriesCompensationAuthority/data-capture-service/pull/199#discussion_r644903934

File: questionnaire/submissions.test.js
Comment (armoj) [reply questionnaire/submissions.test.js:59]: this is just to force an error when a specific id is used. i.e. emulating a failure in a specific scenario. I will adjust it to be async as that was an error

**Review by sinclairs — CHANGES_REQUESTED** [review]
Probs best talking through the rest of this review. I'll see if you're available on slack.

File: questionnaire/submissions.test.js
Comment (sinclairs) [inline questionnaire/submissions.test.js:519]: This test isn't independent. These lines could be removed and it would still pass due to the previous test

File: questionnaire/submissions.test.js
Comment (sinclairs) [inline questionnaire/submissions.test.js:533]: This doesn't appear to assert "should resubmit failed submissions". Given you're returning from `postFailedSubmissions`, why not return that in the response and check that it's as expected?

**Review by sinclairs — CHANGES_REQUESTED** [review]

File: questionnaire/submissions.test.js
Comment (armoj) [reply questionnaire/submissions.test.js:528]: that is what I originally had and that was said to be not a good approach. So I removed a "useful" response.

Comment (armoj) [reply questionnaire/submissions.test.js:528]: I can add it back in...

File: questionnaire/submissions.test.js
Comment (armoj) [reply questionnaire/submissions.test.js:505]: This test is to ensure that [this](https://github.com/CriminalInjuriesCompensationAuthority/data-capture-service/blob/add-resubmission-function/questionnaire/submissions-service.js#L22) line was hit, BUT not [this](https://github.com/CriminalInjuriesCompensationAuthority/data-capture-service/blob/add-resubmission-function/questionnaire/submissions-service.js#L29) line

File: questionnaire/submissions-routes.js
Comment (sinclairs) [inline questionnaire/submissions-routes.js:19]: Can you remove this as it no longer applies

**Review by sinclairs — CHANGES_REQUESTED** [review]
Just a comment to remove.

Given the changes do you want to check on UAT before approval?

**Comment by armoj** [comment]
> Just a comment to remove.
> 
> Given the changes do you want to check on UAT before approval?

**Review by sinclairs — APPROVED** [review]
Mind and check that this assumption is correct before deploying: https://github.com/CriminalInjuriesCompensationAuthority/data-capture-service/blob/bdf1c6fd6c83c7a4255a23ab6dd74d10273678a3/questionnaire/submissions-service.js#L23

**Pull request merged**
